x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/autoscaling/NodeCpuRatio.java

```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.ml.autoscaling;
+
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.xpack.core.ml.MlMetadata;
+
+import java.util.OptionalDouble;
+
+import static org.elasticsearch.xpack.ml.MachineLearning.CPU_RATIO_NODE_ATTR;
+
+/**
+ * Returns the node CPU taking into account node attributes and cluster state
+ */
+public final class NodeCpuRatio {
+
+    /**
+     * @param state current cluster state
+     * @param node Node whose ratio to return
+     * @return The current ration if found
+     */
+    public static OptionalDouble nodeCpuRatio(ClusterState state, DiscoveryNode node) {
+        String ratioAttr = node.getAttributes().get(CPU_RATIO_NODE_ATTR);
+        if (ratioAttr != null) {
+            try {
+                return OptionalDouble.of(Double.parseDouble(ratioAttr));
+            } catch (NumberFormatException ex) {
+                assert false : "unable to parse ml.memory_cpu_ratio attribute [" + ratioAttr + "] " + ex.getMessage();
+                return OptionalDouble.empty();
+            }
+        }
+        double storedRatio = MlMetadata.getMlMetadata(state).getCpuRatio();
+        return storedRatio > 0 ? OptionalDouble.of(storedRatio) : OptionalDouble.empty();
```

Comment:
I wonder if we should return the higher of the two choices?

If a cluster has scaled up and then back down again then this is going to return numbers based on a smaller node size in preference to the number we learnt at the highest scale ever seen.  Since we're going to use this number for subsequent scaling up the ratio learnt at the biggest size is more reflective of the actual ratio.

The numbers obtained from smaller nodes are more distorted by CPU boosting at small scale and the effect of Filebeat and Metricbeat on available memory.

Reply:
`CPU_RATIO_NODE_ATTR` is calculated and accurate for the tier and is eventually correct.

I will have to think about using the "max". The tricky thing here is that we COULD be scaling up in such a way that the future scale forces `CPU_RATIO_NODE_ATTR` to be exactly accurate, while our "guess" is completely wrong for the future tier size. 

My gut is always use `CPU_RATIO_NODE_ATTR` if its there.